File: frontend/src/api/waitlist.ts
import { apiFetch, type SessionLike } from "./client";
import type { Reservation } from "./reservations";

export type WaitlistStatus = "waiting" | "promoted" | "cancelled" | "expired";

export type WaitlistEntry = {
  id: number;
  hotel_id: number;
  guest_id: number;
  category_id: number;
  promoted_reservation_id?: number | null;
  check_in_date: string;
  check_out_date: string;
  num_adults: number;
  num_children: number;
  status: WaitlistStatus;
  priority: number;
  notes?: string | null;
  contact_phone?: string | null;
  created_at: string;
  promoted_at?: string | null;
  expired_at?: string | null;
};

export type WaitlistEntryCreate = {
  guest_id: number;
  category_id: number;
  check_in_date: string;
  check_out_date: string;
  num_adults?: number;
  num_children?: number;
  priority?: number;
  notes?: string | null;
  contact_phone?: string | null;
};

export type WaitlistPromotePayload = {
  room_id?: number | null;
  notes?: string | null;
};

export type WaitlistPromoteResponse = {
  waitlist_entry: WaitlistEntry;
  reservation: Reservation;
};

export const listWaitlistEntries = (statusFilter = "", session?: SessionLike) => {
  const query = statusFilter ? `?status_filter=${encodeURIComponent(statusFilter)}` : "";
  return apiFetch<WaitlistEntry[]>(`/api/waitlist/${query}`, { session });
};

export const createWaitlistEntry = (payload: WaitlistEntryCreate, session?: SessionLike) =>
  apiFetch<WaitlistEntry>("/api/waitlist/", { method: "POST", data: payload, session });

export const promoteWaitlistEntry = (entryId: number, payload: WaitlistPromotePayload, session?: SessionLike) =>
  apiFetch<WaitlistPromoteResponse>(`/api/waitlist/${entryId}/promote`, {
    method: "POST",
    data: payload,
    session
  });

export const cancelWaitlistEntry = (entryId: number, session?: SessionLike) =>
  apiFetch<WaitlistEntry>(`/api/waitlist/${entryId}/cancel`, { method: "POST", session });

File: app/schemas/waitlist.py
"""Schemas for hotel-scoped waitlist entries."""
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field

from app.models.waitlist import WaitlistStatusEnum
from app.schemas.reservation import ReservationRead


class WaitlistEntryCreate(BaseModel):
    guest_id: int
    category_id: int
    check_in_date: date
    check_out_date: date
    num_adults: int = Field(default=1, gt=0)
    num_children: int = Field(default=0, ge=0)
    priority: int = Field(default=100, ge=0)
    notes: Optional[str] = None
    contact_phone: Optional[str] = Field(default=None, max_length=50)


class WaitlistEntryUpdate(BaseModel):
    priority: Optional[int] = Field(default=None, ge=0)
    notes: Optional[str] = None
    contact_phone: Optional[str] = Field(default=None, max_length=50)


class WaitlistEntryRead(BaseModel):
    id: int
    hotel_id: int
    guest_id: int
    category_id: int
    promoted_reservation_id: Optional[int] = None
    check_in_date: date
    check_out_date: date
    num_adults: int
    num_children: int
    status: WaitlistStatusEnum
    priority: int
    notes: Optional[str] = None
    contact_phone: Optional[str] = None
    created_at: datetime
    promoted_at: Optional[datetime] = None
    expired_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class WaitlistPromoteRequest(BaseModel):
    room_id: Optional[int] = None
    notes: Optional[str] = None


class WaitlistPromoteResponse(BaseModel):
    waitlist_entry: WaitlistEntryRead
    reservation: ReservationRead
